tests/JobDefinitionFactoryFunctionalTest.php

```diff
@@ -0,0 +1,266 @@
+<?php
+
+declare(strict_types=1);
+
+namespace App\Tests;
+
+use App\JobDefinitionFactory;
+use Keboola\ConfigurationVariablesResolver\ComponentsClientHelper;
+use Keboola\DockerBundle\Docker\Component;
+use Keboola\DockerBundle\Docker\JobDefinition;
+use Keboola\JobQueueInternalClient\JobFactory\Job;
+use Keboola\JobQueueInternalClient\JobFactory\JobInterface;
+use Keboola\JobQueueInternalClient\JobFactory\ObjectEncryptor\JobObjectEncryptor;
+use Keboola\PermissionChecker\BranchType;
+use Keboola\StorageApi\ClientException;
+use Keboola\StorageApi\Components;
+use Keboola\StorageApi\DevBranches;
+use Keboola\StorageApi\Options\Components\Configuration as StorageConfiguration;
+use Keboola\StorageApi\Options\Components\ConfigurationRow;
+use Keboola\StorageApiBranch\ClientWrapper;
+use Keboola\StorageApiBranch\Factory\ClientOptions;
+use Keboola\StorageApiBranch\Factory\StorageClientPlainFactory;
+use Keboola\VaultApiClient\Variables\Model\ListOptions;
+use Keboola\VaultApiClient\Variables\VariablesApiClient;
+use Symfony\Bundle\FrameworkBundle\Test\KernelTestCase;
+
+class JobDefinitionFactoryFunctionalTest extends KernelTestCase
```

Comment:
Tady v tom jsem chtel jeste udelat test, ze to spravne rozpadne rows na jobDefinition a replacne v nich variables, ale zacal jsem se v tom celkem zamotavat: variables_values_id muze byt z hlavniho cofingu, muze byt z row, muze se to nejak kombinovat treba s branchema atd.

Hodil by se mi asi nejaky nazor z venku jestli tam jeste neco takove patlat, pripadne jak to prekombinovavat.

Reply:
No variable_values_id  hlavního configu a současně z row být nemůže - jde to z jobu a jak se to tam dostalo ti tady může být jedno (ale z row to není https://github.com/keboola/job-queue/blob/2f8f2c6c1e3515dd4076db41d0c18465f8edd599/libs/internal-api-php-client/src/JobFactory/JobRuntimeResolver.php#L108), v každým případě, v tuto chvíli už je to statická hodnota.

To, že se variables replacnou otestovaný máš a to, že se to správně rozpadne je otastovaný v docker bundle.
https://github.com/keboola/docker-bundle/blob/master/Tests/Docker/JobDefinitionParserTest.php, takže imho je to celý dobrý.

Jedna věc tu ale chybí imho - otestovat, že se replacnou variables v shared codu - je to dobře, ale otestovaný to není a reálně se to používá.

Reply:
:+1: doplnil jsme test, ze se replacnou shared codes a i variables v nich